File: src/104.maximum_depth_of_binary_tree.kt
fun main() {
    println(maxDepth(
        TreeNode(3).apply {
            left = TreeNode(9)
            right = TreeNode(20).apply {
                left = TreeNode(15)
                right = TreeNode(7)
            }
        }
    ))

    println(maxDepth(
        TreeNode(1).apply {
            right = TreeNode(2)
        }
    ))
}

fun maxDepth(root: TreeNode?): Int {
    return when {
        root == null -> 0
        root.left == null && root.right == null -> 1
        else -> Math.max(maxDepth(root.left), maxDepth(root.right)) + 1
    }
}

File: lib/104.maximum_depth_of_binary_tree.dart
import 'dart:math';

import 'other/tree_node.dart';

void main() {
  print(maxDepth(
      TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))));
  print(maxDepth(TreeNode(1, null, TreeNode(2))));
}

int maxDepth(TreeNode? root) {
  if (root == null) {
    return 0;
  } else if (root.left == null && root.right == null) {
    return 1;
  } else {
    return max(maxDepth(root.left), maxDepth(root.right)) + 1;
  }
}
